### ISSUE TITLE
More math functions: ceil, floor, log, log2

### ISSUE THREAD

**Issue by KOLANICH** [kaitai-io/kaitai_struct#232]

**Comment by GreyCat** [comment]
Any particular reasons to have them?

**Comment by KOLANICH** [comment]
Calculation of count of bits needed to store a number up to known maximum value.

**Comment by GreyCat** [comment]
I just wonder if anyone uses them in real-life parsers (probably looking up one of several values via a simple decision tree would be faster)?

Technically, x86 FPU indeed has FYL2X and FYL2XP1...

**Comment by KOLANICH** [comment]
https://github.com/nongeneric/lsd2dsl/blob/master/dictlsd/LenTable.cpp#L82
https://github.com/nongeneric/lsd2dsl/blob/master/dictlsd/tools.cpp#L30
```BitLength(count)``` is essentially ```floor(log2(count))+1```.

**Comment by GreyCat** [comment]
Yeah, but I suspect that their implementation is much faster than using FPU...

**Comment by KOLANICH** [comment]
Yes. That's why I also insist on dedicated function for that.

**Comment by GreyCat** [comment]
Just tried a simple test, i.e.:

```c
unsigned BitLength1(unsigned num) {
    unsigned res = 1;
    while ((num >>= 1) != 0)
        res++;
    return res;
}

unsigned BitLength2(unsigned num) {
        return floor(log2(num)) + 1;
}
```

1KKK iterations on my laptop take:

* ~10.95s using loop version
* ~20.40s using FPU version

Both are using basic `gcc -O2` => FPU is ~1.8x slower.

**Comment by KOLANICH** [comment]
BSR doesn't use fpu. Use intrinsics for your compiler.

**Comment by GreyCat** [comment]
Thanks, I had no idea! :)
ffs is actually compiled into BSF, not BSR:

```
│       ┌─< 0x00400535      eb12           jmp 0x400549
        │   0x00400537      660f1f840000.  nop word [rax + rax]
│      ┌──> 0x00400540      0fbcc2         bsf eax, edx
│      ││   0x00400543      0f44c1         cmove eax, ecx
│      ││   0x00400546      83c001         add eax, 1
│      ││   ; JMP XREF from 0x00400535 (sym.main)
│      │└─> 0x00400549      83c001         add eax, 1
│      │    0x0040054c      83c201         add edx, 1
│      │    0x0040054f      4898           cdqe
│      │    0x00400551      4801c6         add rsi, rax                ; '#'
│      │    0x00400554      81fa00e1f505   cmp edx, 0x5f5e100
│      └──< 0x0040055a      75e4           jne 0x400540
```

It is roughly 4x faster that loop implementation: ~2.70s vs 10.95s.

**Comment by KOLANICH** [comment]
oh, this means that __builtin_ffs is incorrect here. I googled better, there is no intrinsic for bsr in gcc. So for gcc/clang we need ```_bit_scan_reverse```.

**Comment by N3X15** [comment]
> Any particular reasons to have them?

Calculating correct number of bytes to feed to a subtype.  

For example:  ECMA 335 `#US` stream uses UTF-16-encoded strings, so the decoder has to be fed an even number of bytes.  However, the length field provides the count of bytes in the string, PLUS a possible termination byte.

So, the schema ends up looking something like:
```yaml
seq:
- id: length
  type: ms_vlq # Unsigned VLQ defined in ECMA 335 II.24.2.4
- id: string
  type: string
  encoding: utf-16-le
  size: floor(length.value / 2) * 2
- if: (length.value & 1) != 0
  id: terminator
  type: u1
```

Unfortunately, floor() is not available, so I'm having to patch the generated code after the fact (Just replacing the / operator with the Python // operator, in my case).

Hope to see some movement on this.

**Comment by N3X15** [comment]
Update: Out of curiousity, I tried to see if typecasting worked, and it did:
```yaml
   - id: string
      type: str
      encoding: UTF-16-LE
      size: (length_data.value / 2).as<u4> * 2
```
```python
self.string = (self._io.read_bytes((self.length_data.value // 2 * 2))).decode(u"UTF-16-LE")
```
Hilarious.

All jokes aside, this is a hacky solution, IMHO, and Kaitai should provide maximum flexibility where warranted.
